drivers/watchdog/wdt_fwdgt_gd32.c

```diff
@@ -0,0 +1,161 @@
+
+#define FWDGT_DATA(dev) ((struct gd32_fwdgt_data *const)(dev)->data)
+
+/* mutable driver data */
+struct gd32_fwdgt_data {
+	/* peripheral clock frequency*/
+	uint32_t frequency;
+};
+
+/**
+ * @brief Calculates FWDGT config value from timeout.
+ *
+ * @param frequency peripheral clock frequency value.
+ * @param timeout Timeout value in milliseconds.
+ * @param prescaler Pointer to the storage of prescaler value.
+ * @param reload Pointer to the storage of reload value.
+ *
+ * @return 0 on success, -EINVAL if the timeout is out of range
+ */
+static int gd32_fwdgt_convert_timeout(uint32_t frequency,
+				      uint32_t timeout,
+				      uint32_t *prescaler,
+				      uint32_t *reload)
+{
+	uint16_t divider = 4U;
+	uint8_t shift = 0U;
+
+	uint32_t ticks = (uint64_t)frequency * timeout / MSEC_PER_SEC;
+
+	while ((ticks / divider) > FWDGT_RELOAD_MAX) {
+		shift++;
+		divider = 4U << shift;
+	}
+
+	if (!IS_VALID_FWDGT_PRESCALER(PSC_PSC(shift)) || timeout == 0) {
+		return -EINVAL;
+	}
+
+	/* convert the 'shift' to prescaler value */
+	*prescaler = PSC_PSC(shift);
+	*reload = (ticks / divider) - 1U;
+
+	return 0;
+}
+
+static int gd32_fwdgt_setup(const struct device *dev, uint8_t options)
+{
+	ARG_UNUSED(dev);
+
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+		dbg_periph_enable(DBG_FWDGT_HOLD);
+	}
+
+	if (options & WDT_OPT_PAUSE_IN_SLEEP) {
+		return -ENOTSUP;
+	}
+
+	fwdgt_enable();
+	return 0;
+}
+
+static int gd32_fwdgt_disable(const struct device *dev)
+{
+	/* watchdog cannot be stopped once started */
+	ARG_UNUSED(dev);
+
+	return -EPERM;
+}
+
+static int gd32_fwdgt_install_timeout(const struct device *dev,
+				      const struct wdt_timeout_cfg *config)
+{
+	uint32_t prescaler = 0U;
+	uint32_t reload = 0U;
+	struct gd32_fwdgt_data *data = FWDGT_DATA(dev);
+
+	/* Callback is not supported by FWDGT */
+	if (config->callback != NULL) {
+		return -ENOTSUP;
+	}
+
+	/* Calculate prescaler and reload value from timeout value */
+	if (gd32_fwdgt_convert_timeout(data->frequency, config->window.max, &prescaler, &reload)) {
+		return -EINVAL;
+	}
+
+	/* Configure and run FWDGT */
+	fwdgt_write_enable();
+	if (!fwdgt_config(reload, prescaler)) {
+		return -EINVAL;
+	}
+	fwdgt_write_disable();
+
+	return 0;
+}
+
+static int gd32_fwdgt_feed(const struct device *dev, int channel_id)
+{
+	ARG_UNUSED(channel_id);
+	fwdgt_counter_reload();
+	return 0;
+}
+
+static const struct wdt_driver_api fwdgt_gd32_api = {
+	.setup = gd32_fwdgt_setup,
+	.disable = gd32_fwdgt_disable,
+	.install_timeout = gd32_fwdgt_install_timeout,
+	.feed = gd32_fwdgt_feed,
+};
+
+static int gd32_fwdgt_init(const struct device *dev)
+{
+	int ret = 0;
+
+	/* Turn on and wait stabilize oscillator.                  */
+	rcu_osci_on(DT_PROP(FWDGT_NODE, rcu_oscillator));
+	while (!rcu_osci_stab_wait(DT_PROP(FWDGT_NODE, rcu_oscillator))) {
+	}
+	/* TODO: It might better to reimplement with clock_control */
+	FWDGT_DATA(dev)->frequency = DT_PROP(FWDGT_NODE, frequency);
```

Comment:
Is the problem that `clock_control` is not yet implemented for GD32? In general I prefer not to have code with `TODO` comments.

Reply:
I confirmed hardware specs.
Clock source configuration of the watchdogs are same in all GD32 series.
This line no need consideration for modifiable configuration.
I rewrite simply this line with using fixed value and remove TODO.